test/unit/math/prim/prob/hmm_marginal_test.cpp

```diff
@@ -0,0 +1,243 @@
+ * the last element using the fact each column sums to 1.
+ * The purpose of this function is to do finite diff benchmarking,
+ * without breaking the simplex constraint.
+ */
+template <typename T_omega, typename T_Gamma, typename T_rho>
+inline stan::return_type_t<T_omega, T_Gamma, T_rho> hmm_marginal_test_wrapper(
+    const Eigen::Matrix<T_omega, Eigen::Dynamic, Eigen::Dynamic>& log_omegas,
+    const Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>&
+        Gamma_unconstrained,
+    const std::vector<T_rho>& rho_unconstrained) {
+  using stan::math::row;
+  using stan::math::sum;
+  int n_states = log_omegas.rows();
+
+  Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic> Gamma(n_states,
+                                                               n_states);
+  for (int i = 0; i < n_states; i++) {
+    Gamma(i, n_states - 1) = 1 - sum(row(Gamma_unconstrained, i + 1));
+    for (int j = 0; j < n_states - 1; j++) {
+      Gamma(i, j) = Gamma_unconstrained(i, j);
+    }
+  }
+
+  Eigen::Matrix<T_rho, Eigen::Dynamic, 1> rho(n_states);
+  rho(1) = 1 - sum(rho_unconstrained);
+  for (int i = 0; i < n_states - 1; i++)
+    rho(i) = rho_unconstrained[i];
+
+  return stan::math::hmm_marginal_lpdf(log_omegas, Gamma, rho);
+}
+
+/**
+ * In the proposed example, the latent sate x determines
+ * the observational distribution:
+ *  (i) normal(mu, sigma)
+ *  (ii) normal(-mu, sigma)
+ */
+double state_lpdf(double y, double abs_mu, double sigma, int state) {
+  int x = (-2 * state + 1);
+  double chi = (y - x * abs_mu) / sigma;
+  return -0.5 * chi * chi - 0.5 * std::log(6.283185307179586) - std::log(sigma);
+}
+
+class hmm_marginal_lpdf_test : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    n_states = 2;
+    p1_init = 0.65;
+    gamma1 = 0.7;
+    gamma2 = 0.45;
+    n_transitions = 10;
+    abs_mu = 1;
+    sigma = 1;
+
+    Eigen::VectorXd rho_(n_states);
+    rho_ << p1_init, 1 - p1_init;
+    rho = rho_;
+
+    Eigen::MatrixXd Gamma_(n_states, n_states);
+    Gamma_ << gamma1, 1 - gamma1, gamma2, 1 - gamma2;
+    Gamma = Gamma_;
+
+    Eigen::VectorXd obs_data_(n_transitions + 1);
+    obs_data_ << -0.3315914, -0.1655340, -0.7984021, 0.2364608, -0.4489722,
+        2.1831438, -1.4778675, 0.8717423, -1.0370874, 0.1370296, 1.9786208;
+    obs_data = obs_data_;
+
+    Eigen::MatrixXd log_omegas_(n_states, n_transitions + 1);
+    for (int n = 0; n < n_transitions + 1; n++) {
+      log_omegas_.col(n)[0] = state_lpdf(obs_data[n], abs_mu, sigma, 0);
+      log_omegas_.col(n)[1] = state_lpdf(obs_data[n], abs_mu, sigma, 1);
+    }
+    log_omegas = log_omegas_;
+    log_omegas_zero = log_omegas.block(0, 0, n_states, 1);
+
+    std::vector<double> rho_unconstrained_(n_states - 1);
+    for (int i = 0; i < rho.size() - 1; i++)
+      rho_unconstrained_[i] = rho(i);
+    rho_unconstrained = rho_unconstrained_;
+
+    Gamma_unconstrained = Gamma.block(0, 0, n_states, n_states - 1);
+  }
+
+  int n_states, n_transitions;
```

Comment:
There should be `_` suffixes on the member variables (the ones defined as part of the class) and not on local variables.  The convention is backwards here.